### ISSUE TITLE
Further Crunching On Key Level

### ISSUE THREAD

**Issue by johntran** [banterfm/graphql-crunch#1]
Hey there!

I noticed that the values are given numerical values, but not keys. There seems to be a lot of space savings on that front too.

What was the reason for not applying the same optimizations to keys? GraphQL properties can only be alphabetical characters, so we know that the keys can never be numbers.

Thanks for open sourcing this!

**Comment by jamesreggio** [comment]
Good observation. I think it was, in fact, an oversight 😄

We'd love to take a PR for this, if you're willing to submit one.

My only fear would be that the JS VMs interpret an object with all numeric keys as a sparse array and thus de-optimize. We could conceivably use an array of pairs to work around this issue, but we'd have to tag it to differentiate it from an actual array. (Let me know if this isn't clear.)

**Comment by kachkaev** [comment]
@jamesreggio are you planning do address this issue in the near future and release `graphql-crunch@2.0` that also crunches keys? If so, it might be worth waiting a bit with https://github.com/graphcool/graphql-yoga/issues/303 and https://github.com/banterfm/graphql-crunch/pull/3 to avoid downstream breaking changes.

cc @lfades

**Comment by kachkaev** [comment]
Speaking of the issue with keys that you are mentioning. When you're referring to _tagging_ key-value -pairs, do you mean that this needs to be done internally in `uncrunch`? If so, uncrunched arrays could simply go to a second temporary array that's pre-filled with nulls and this trick may work like this _tagging_.

Alternatively, object keys could be prefixed like `{'k42': 1, 'k142': 67 }` to avoid turning them into sparse arrays. That's one extra character and one extra `parseInt(string.slice())` operation per property. However, it's hard for me to tell in advance what's more performant without actual benchmarking 🤔 

In any case, crunching keys feels like a thing worth a try!

**Comment by kachkaev** [comment]
@stevekrenzel what's the performance regression in both cases? (just curious)

Do you think there is chance that key crunching will be released in a near future? It'd be great to have some knowledge on this to decide when to proceed with with https://github.com/graphcool/graphql-yoga/issues/303.

**Comment by williambailey** [comment]
Would it be possible to offer this functionality in a why where the library exposes the different crunch algorithms so that consumers can choose how they want their data crunched.

This would allow developers to build servers that allow clients to specify the version they understand while keeping potential for data changing formats in the future. It would also allow for client language optimisations where `{"k1":2,"k3":4}` might pre preferred for a js client (should sparse arrays be an issue) while `{"1":2,"3":4}` might be more natural for non js clients.

It might then be worth recommending that servers add a response extension that clients can inspect to see if the crunch request we honored. This would allow servers to drop crunch support without impacting consumers.

Putting the two together a `formatResponse` might looks something like...

```js
app.use('/graphql', graphqlExpress((request) => {
return {
  formatResponse: (response) => {
    const crunch = request.query.crunch; /* or from a header */
    if(crunch && response.data) {
      response.extensions = response.extensions || {};
      r.extensions.crunch = { version: crunch };
      switch(crunch) {
        case 1:
          r.data = crunch(r.data);
          break;
        // case 2 is not something this server wants to support.
        case 3:
          r.data = crunch3(r.data);
          break;
        default:
          delete r.extensions.crunch;
          break;
      }
    }
    return response;
  }
};
}));
```

**Comment by kachkaev** [comment]
Hey folks 👋 

Are you still considering this improvement or did it demonstrate itself as not worth it? Curious to hear about your experience.

**Comment by stevekrenzel** [comment]
Hey, so I can finally and definitively reply to this thread. 

Deduplicating keys resulted in smaller raw payloads, but the gzipped sizes were larger and crunching took longer. Gzip handles recurring primitives like a champ, so in GraphQL-Crunch 2.0 we actually took the opposite approach. We only deduplicate containers (arrays and objects). This often results in payloads that are 20%-25% smaller than GraphlQL-Crunch 1.0 after begin gzipped. In a future update to the README I'll include the benchmark results.

The new update also crunches 2x-3x faster. It's blazing fast.

@williambailey, 2.0 was implemented to still support v1. The client can ask for a version and if the server doesn't know the version number then it will default to 1.0. I've updated the README to show how this can work.

Thanks all for the continued feedback! I'm going close this thread now.
